File: island.cpp
class Solution {
public:
vector<vector<char>> grid;
vector<vector<int>> dirs={{0,1},{1,0},{0,-1},{-1,0}};
vector<vector<bool>> seen;
int count=0;
int n, m;
    int numIslands(vector<vector<char>>& grid) {
        this->grid=grid;
         n=grid.size();
         m=grid[0].size();

        seen=vector(n,vector<bool>(m,false));

        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(grid[i][j]=='1' and !seen[i][j])
                {
                    seen[i][j]=true;
                    dfs(i,j);
                    count++;
                }
            }
        }
        return count;
    }

    void dfs(int r, int c)
    {
        for(vector<int> d:dirs)
        {
            int newR=r+d[0];
            int newC=c+d[1];
            if(valid(newR, newC) and !seen[newR][newC])
            {
                seen[newR][newC]=true;
                dfs(newR, newC);
            }
        }
    }
    bool valid(int r, int c)
    {
        return r>=0 and c>=0 and r<n and c<m and grid[r][c]=='1';
    }
};

File: island.py

from collections import deque

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def bfs(r, c):
            if r < 0 or r >= row or c < 0 or c >= col or visited[r][c] or grid[r][c] == "0":
                return 0
            visited[r][c] = True
            directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                bfs(nr, nc)
            return 1

        if not grid or not grid[0]:
            return 0

        row, col = len(grid), len(grid[0])
        visited = [[False] * col for _ in range(row)]
        count = 0

        for i in range(row):
            for j in range(col):
                if not visited[i][j] and grid[i][j] == "1":
                    count += bfs(i, j)

        return count
